logic/LogicManager.java

```diff
@@ -18,43 +18,133 @@
 
 package grakn.core.logic;
 
+import grakn.core.common.exception.GraknException;
+import grakn.core.common.iterator.ResourceIterator;
+import grakn.core.common.parameters.Label;
 import grakn.core.concept.ConceptManager;
+import grakn.core.concept.type.RelationType;
 import grakn.core.graph.GraphManager;
 import grakn.core.graph.structure.RuleStructure;
+import grakn.core.graph.util.Encoding;
 import grakn.core.logic.tool.TypeHinter;
+import grakn.core.traversal.TraversalEngine;
+import graql.lang.pattern.Conjunctable;
 import graql.lang.pattern.Conjunction;
+import graql.lang.pattern.Negation;
 import graql.lang.pattern.Pattern;
+import graql.lang.pattern.constraint.TypeConstraint;
+import graql.lang.pattern.variable.BoundVariable;
 import graql.lang.pattern.variable.ThingVariable;
 
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static grakn.core.common.exception.ErrorMessage.RuleWrite.TYPES_NOT_FOUND;
+
 public class LogicManager {
 
     private final ConceptManager conceptMgr;
     private final GraphManager graphMgr;
     private final TypeHinter typeHinter;
+    private LogicCache logicCache;
 
-    public LogicManager(GraphManager graphMgr, ConceptManager conceptMgr, TypeHinter typeHinter) {
+    public LogicManager(GraphManager graphMgr, ConceptManager conceptMgr, TraversalEngine traversalEng, LogicCache logicCache) {
         this.graphMgr = graphMgr;
         this.conceptMgr = conceptMgr;
-        this.typeHinter = typeHinter;
+        this.logicCache = logicCache;
+        this.typeHinter = new TypeHinter(conceptMgr, traversalEng, logicCache.hinter());
     }
 
     public Rule putRule(String label, Conjunction<? extends Pattern> when, ThingVariable<?> then) {
-        final RuleStructure structure = graphMgr.schema().getRule(label);
-        if (structure != null) structure.delete();
-        Rule rule = Rule.of(conceptMgr, graphMgr, typeHinter, label, when, then);
-
-        // TODO detect negated cycles in the rule graph after inserting this rule and updating a rule graph
-
-        return rule;
+        RuleStructure structure = graphMgr.schema().getRule(label);
+        if (structure != null) {
+            // overwriting a rule means we purge it and re-create the rule
+            structure.delete();
+            logicCache.rule().invalidate(label);
+        }
+        return logicCache.rule().get(label, l -> Rule.of(graphMgr, conceptMgr, this, label, when, then));
```

Comment:
We have ensured the rule has been deleted if it exists; So this last line is creating a rule, putting it in the cache, and storing it to the graph, and then returning it from the cache?

Reply:
yup - and atomically!